src/validation/rules/SingleFieldSubscriptionsRule.js

```diff
@@ -1,28 +1,119 @@
 import { GraphQLError } from '../../error/GraphQLError';
 
 import type { ASTVisitor } from '../../language/visitor';
-import type { OperationDefinitionNode } from '../../language/ast';
+import type {
+  OperationDefinitionNode,
+  SelectionSetNode,
+} from '../../language/ast';
+import { Kind } from '../../language/kinds';
 
 import type { ASTValidationContext } from '../ValidationContext';
 
 /**
- * Subscriptions must only include one field.
+ * Walks the selection set and returns a list of selections where extra fields
+ * were selected, and selections where introspection fields were selected.
+ */
+function walkSubscriptionSelectionSet(
```

Comment:
Any reason why this can't be a more typical `CollectFields()` implementation? I'm surprised there isn't something to be reused from another validation rule.

Reply:
The closest I could find was `_collectFieldsAndFragmentNames` from `src/validation/rules/OverlappingFieldsCanBeMergedRule.js` but it doesn't recurse into named fragments. We also can't use collectFields from execution because it requires an execution context.

My algorithm is more efficient than recursing over calls to `_collectFieldsAndFragmentNames` because it has no need to reference the underlying types - there's no need for calls to `parentType.getFields()`/`typeFromAST`/`isObjectType`/etc since all we want to know is:

1. is there any `__introspection` fields?
2. is there more than one field?

The actual type they're defined on (or even whether it exists or not!) is irrelevant.

Reply:
I suppose that could potentially be untrue; e.g.

```graphql
type Query implements Node {...}
type Subscription implements Node {...}

subscription ShouldBeFine($bool: Boolean!) {
  ... on Node {
    ... on Query { __typename id firstName }
    ... on Subscription {
      mySubscriptionField @include(if: $bool)
      myOtherSubscriptionField @skip(if: $bool)
    }
  }
}
```

in this case a runtime `collectFields` would return only `mySubscriptionField` or `myOtherSubscription` field (depending on the value of `$bool`) because the Query selections would be scoped out as incompatible types and the `@include`/`@skip` would ensure only one field was valid. This is problematic though, since in validate we don't have access to the runtime variables so we cannot determine which of the fields will/won't be included.

In the [spec edit](https://github.com/graphql/graphql-spec/pull/776/files) I wrote this leveraging the `CollectFields` call which was already present, which already relies on `varibleValues`; but GraphQL.js allows validation without variables so... is this a fundamental issue in the way GraphQL.js is built? Personally I think we should edit the spec to reflect GraphQL.js' behaviour - validating the query without access to the variables seems desirable and so having the validation rule for "single root field" depend on CollectFields which depends on variableValues is undesirable.

Reply:
Reading more carefully the spec explicitly makes variableValues an empty object, so in this case the operation would be valid; however so would the following:

```graphql
type Query implements Node {...}
type Subscription implements Node {...}

subscription ShouldBeFine($bool: Boolean!) {
  ... on Node {
    ... on Query { __typename id firstName }
    ... on Subscription {
      mySubscriptionField1 @include(if: $bool)
      mySubscriptionField2 @include(if: $bool)
      mySubscriptionField3 @skip(if: $bool)
    }
  }
}
```

In this case the CollectFields algorithm would result in only `mySubscriptionField3` being considered; the operation would be valid, but then if `$bool` was set to `true` at runtime there'd be two selections and the operation would be invalid.

This is definitely a spec bug. I'm going to write up a proper fix.

Reply:
(After more thought, I'm not sure it's a bug so much as an "unexpected solution"; but I've filed a potential spec edit anyway: https://github.com/graphql/graphql-spec/pull/860)

Given the above discussion I've rewritten the solution to use CollectFields so that it can take into account the fragment compatibility and the `@include`/`@skip` directives in the same way that the GraphQL spec would.